src/store/utils.ts

```diff
@@ -0,0 +1,95 @@
+export interface ErrorInfo {
+  message: string;
+}
+
+// Represents a piece of remote info, which may be
+// - No info and not loading (initial)
+// - No info and loading (fetching)
+// - Has info and not loading (completed)
+// - Has info and loading (refreshing)
+export type Remote<T> = { result?: T; loading: boolean };
+
+/* Represents result of a task, either success (with data) or error. */
+export type Result<T, E = ErrorInfo> = DataResult<T> | ErrorResult<E>;
+export type DataResult<T> = { data: T };
+export type ErrorResult<E> = { error: E };
+
+export type RemoteResult<T, E = ErrorInfo> = Remote<Result<T, E>>;
+
+const ownProp = Object.prototype.hasOwnProperty;
+
+export function hasData<T, E>(
+  result: Result<T, E> | undefined
+): result is DataResult<T> {
+  return result !== undefined && ownProp.call(result, 'data');
```

Comment:
I think you want to use 'in' here and below, it's a little more succinct

https://www.typescriptlang.org/docs/handbook/advanced-types.html#using-the-in-operator

Reply:
Updated. People back in the days tell me not to use `in` because it walks through the whole prototype chain, but whatever.